File: src/config/cli_paths.cpp
#include "config/cli_paths.h"

#include <cstdlib>
#include <string>

#if defined(_WIN32)
#ifndef NOMINMAX
#define NOMINMAX
#endif
#include <cstring>

#include <windows.h>
#endif

#include "rac/desktop/rac_desktop.h"

namespace wally::paths {

namespace {

// Read an environment variable as UTF-8. On Windows std::getenv decodes the
// value through the process ANSI code page, corrupting Unicode paths (e.g. an
// international LOCALAPPDATA/USERPROFILE/RUNANYWHERE_HOME); read the wide value
// and convert with CP_UTF8 instead. Returns empty when unset or empty.
std::string getenv_utf8(const char* name) {
#if defined(_WIN32)
    const std::wstring wname(name, name + std::strlen(name));  // name is ASCII
    const wchar_t* wvalue = _wgetenv(wname.c_str());
    if (!wvalue || wvalue[0] == L'\0') {
        return {};
    }
    const int len = WideCharToMultiByte(CP_UTF8, 0, wvalue, -1, nullptr, 0, nullptr, nullptr);
    if (len <= 1) {
        return {};
    }
    std::string out(static_cast<size_t>(len - 1), '\0');  // len counts the NUL
    WideCharToMultiByte(CP_UTF8, 0, wvalue, -1, out.data(), len, nullptr, nullptr);
    return out;
#else
    const char* value = std::getenv(name);
    return (value && value[0] != '\0') ? std::string(value) : std::string();
#endif
}

}  // namespace

std::string normalize_dir(std::string dir) {
#if defined(_WIN32)
    // Windows environment values (LOCALAPPDATA, USERPROFILE) come back
    // backslash-separated, while every path built from them here appends
    // '/'-joined segments -- leaving `wally info` printing a mixed
    // C:\Users\...\AppData\Local/RunAnywhere. Fold to '/' so a single
    // style survives into the output; Win32 accepts either separator.
    for (char& c : dir) {
        if (c == '\\') {
            c = '/';
        }
    }
#endif
    while (dir.size() > 1 && (dir.back() == '/' || dir.back() == '\\')) {
        dir.pop_back();
    }
    return dir;
}

std::string resolve_home(const std::string& override_dir) {
    if (!override_dir.empty()) {
        return normalize_dir(override_dir);
    }
    if (std::string env = getenv_utf8("RUNANYWHERE_HOME"); !env.empty()) {
        return normalize_dir(std::move(env));
    }
    char buffer[1024] = {};
    if (rac_desktop_default_base_dir(buffer, sizeof(buffer)) == RAC_SUCCESS) {
        return normalize_dir(buffer);
    }
    return {};
}

std::string state_dir() {
    if (std::string env = getenv_utf8("XDG_STATE_HOME"); !env.empty()) {
        return normalize_dir(std::move(env)) + "/runanywhere";
    }
    if (std::string home = getenv_utf8("HOME"); !home.empty()) {
        return normalize_dir(std::move(home)) + "/.local/state/runanywhere";
    }
#if defined(_WIN32)
    if (std::string local = getenv_utf8("LOCALAPPDATA"); !local.empty()) {
        return normalize_dir(std::move(local)) + "/RunAnywhere/state";
    }
    if (std::string profile = getenv_utf8("USERPROFILE"); !profile.empty()) {
        return normalize_dir(std::move(profile)) + "/AppData/Local/RunAnywhere/state";
    }
#endif
    return {};
}

}  // namespace wally::paths

File: src/config/cli_paths.rs
//! wally directory resolution.
//!
//! One knob controls where models live: the RunAnywhere HOME directory.
//!   resolution: --home flag → $RUNANYWHERE_HOME → ${XDG_DATA_HOME:-~/.local/share}/runanywhere
//! Models are derived BY COMMONS from that home via rac_model_paths_*
//! (home named "runanywhere" → <home>/Models/{framework}/<id>, the same layout
//! the Linux test rig and Playground tooling use).
//!
//! Config (secure store) stays under ${XDG_CONFIG_HOME:-~/.config}/runanywhere;
//! REPL history under ${XDG_STATE_HOME:-~/.local/state}/runanywhere.

use crate::sys;
use crate::util::getenv;

/// Strip trailing '/' (keeps root "/"). On Windows also folds '\\' to '/'.
pub fn normalize_dir(dir: &str) -> String {
    // Windows environment values (LOCALAPPDATA, USERPROFILE) come back
    // backslash-separated, while every path built from them here appends
    // '/'-joined segments -- leaving `wally info` printing a mixed
    // C:\Users\...\AppData\Local/RunAnywhere. Fold to '/' so a single style
    // survives into the output; Win32 accepts either separator.
    #[cfg(windows)]
    let mut dir = dir.replace('\\', "/");
    #[cfg(not(windows))]
    let mut dir = dir.to_string();
    while dir.len() > 1 && (dir.ends_with('/') || dir.ends_with('\\')) {
        dir.pop();
    }
    dir
}

/// Resolve the RunAnywhere home (storage base dir) — see the module header for
/// the precedence. Returns the empty string only when $HOME is unresolvable.
pub fn resolve_home(override_dir: &str) -> String {
    if !override_dir.is_empty() {
        return normalize_dir(override_dir);
    }
    if let Some(env) = getenv("RUNANYWHERE_HOME") {
        return normalize_dir(&env);
    }
    let mut buffer = [0 as std::ffi::c_char; 1024];
    // SAFETY: the kit writes at most `len` bytes, NUL-terminated, into buffer.
    let rc = unsafe { sys::rac_desktop_default_base_dir(buffer.as_mut_ptr(), buffer.len()) };
    if rc == sys::SUCCESS {
        // SAFETY: NUL-terminated by the kit on success.
        let dir = unsafe { std::ffi::CStr::from_ptr(buffer.as_ptr()) };
        return normalize_dir(&dir.to_string_lossy());
    }
    String::new()
}

/// ${XDG_STATE_HOME:-~/.local/state}/runanywhere (not created). On Windows,
/// with XDG_STATE_HOME unset, %LOCALAPPDATA%/RunAnywhere/state even when HOME
/// is set (MSYS2 / Git Bash).
pub fn state_dir() -> String {
    if let Some(env) = getenv("XDG_STATE_HOME") {
        return normalize_dir(&env) + "/runanywhere";
    }
    // Checked ahead of HOME: MSYS2 / Git Bash set HOME on Windows, which would
    // otherwise divert state into %USERPROFILE%/.local/state/runanywhere and
    // leave the two branches below permanently unreachable.
    #[cfg(windows)]
    {
        if let Some(local) = getenv("LOCALAPPDATA") {
            return normalize_dir(&local) + "/RunAnywhere/state";
        }
        if let Some(profile) = getenv("USERPROFILE") {
            return normalize_dir(&profile) + "/AppData/Local/RunAnywhere/state";
        }
    }
    if let Some(home) = getenv("HOME") {
        return normalize_dir(&home) + "/.local/state/runanywhere";
    }
    String::new()
}
